### ISSUE TITLE
Search available images in container registry when deploying from settings

### ISSUE THREAD

**Issue by codegoblins** [kubernetes/dashboard#511]
I'd like to see some integration with container registries to make deploying images easier for the end user. 

The way I see this feature breaks down to 2 tasks:
- Add an endpoint to the backend to fetch a list of public images from a configured container registry endpoint. Ideally supporting multiple container registry providers (we use Quay.io but I know a lot of people use Docker's offering).
- Add a widget to the deploy from settings page to search the offered images and let the user select the desired image. This could be an auto-complete on the container image text box, or a button adjacent to the text box opening a selector modal.

**Comment by floreks** [comment]
We were planning to add autocomplete and some support for docker, and possibly other providers but there were many other things to do and polish before release. I believe this will be implemented at some point. Maybe even in the next release. We'll know more after next planning.

**Comment by bryk** [comment]
@codegoblins Sounds just about right. I think if you have spare cycles, it is ready to be implemented. We've been planning such feature anyways :)

Are you taking this?

**Comment by codegoblins** [comment]
Sure, I will take this.

**Comment by mic4ael** [comment]
Hey, I would like to contribute to the kubernetes-related projects and it feels like starting with this one would a good first issue for me. Is it up for grabs?

**Comment by maciaszczykm** [comment]
@mic4ael Yes. It would be good to create a pull request early so we can discuss the architecture of it. Or you can write that down before you start if you already have an idea.

**Comment by mic4ael** [comment]
Sure! I will first take a closer look on how everything works internally
and then I will come up with some initial implementation plans.

**Comment by maciaszczykm** [comment]
Sounds great.

**Comment by prnam** [comment]
Is there any work already done on this, @floreks / @maciaszczykm? 

Any notes of previous discussions world of great help to us to determine if this is something can be worked from where it was left or need fresh start.

**Comment by maciaszczykm** [comment]
@prnam It would need fresh start.

**Comment by prnam** [comment]
Q1: Is this feature request to store only the registry providers endpoint in settings, then using this endpoint GET list of images (by verified / official / all if allowed by registry provider) while typing in "container image" input filed and caching for sometime? 

Fetching all images on public / private images itself is not helpful in my opinion but the title and description nor I think it is even allowed by private registry at one go or incremental. Please help me understand the request better @maciaszczykm , @codegoblins .  🙂

**Comment by prnam** [comment]
Q2: Do we need to pre-populate a list of well known public image registry endpoint or let the user add themselves somewhere in the settings? 

- If pre-populate, then what are those list of well known registry providers we need to add?

**Comment by prnam** [comment]
Q3: Do we need an images already pulled / built (may date back to very long ago and may have very long list) through any method (docker pull, podman pull, docker build, buildah bud etc.) be recognised and listed on dashboard as in when you type in the “container image” input field?

**Comment by prnam** [comment]
Q4: Do you think this should be an extension / K8s operator to the K8s dashboard / CLI method and not baked within dashboard?

**Comment by haripriya1909** [comment]
i want to work on this
/assign

**Comment by maciaszczykm** [comment]
@haripriya1909 if you want to work on it you need to sync with @prnam as he was first to claim it.

**Comment by maciaszczykm** [comment]
>Q1: Is this feature request to store only the registry providers endpoint in settings, then using the endpoint GET "list of images" (by verified / official / all if allowed by registry provider) while typing in "container image" input filed and caching for sometime?
>
>Fetching all "images" from public / private registry itself is not helpful in my opinion nor I think it is even allowed [rate-limit] by registry at one go or incremental; but the title and description ask is this I believe.
>
>Please help me understand the request better @maciaszczykm , @codegoblins . 🙂

I think there is no real plan for this one yet. We agreed that this would be a nice to thing to have but I didn't spend any time to get more familiar with this topic and provide any proposal. If you want to work on it then a proposal would be a good way to start. Consider some options and provide your insights. We'll try to help. @codegoblins do you have any ideas that might help at this point?

**Comment by maciaszczykm** [comment]
> Q4: Do you think this should be an extension / K8s operator to the K8s dashboard / CLI method and not baked within dashboard?

Something dashboard-specific will be enough unless you have different plans.

**Comment by maciaszczykm** [comment]
> Q2: Do we need to pre-populate a list of well known public image registry endpoint or let the user add themselves somewhere in the settings?

Having settings is always better but can be added later on.

> If pre-populate, then what are those list of well known registry providers we need to add?

Start with something simple.

**Comment by codegoblins** [comment]
Apologies for creating an issue then going dark for 6 years but I stopped working on k8s related projects so I'm afraid I don't have much to add. At the time I was thinking it would be nice to have some form of autocomplete, whether that was searching a public or private registry, or even just remembering images I've deployed before (e.g. in localstorage). That was based on my use of the software in its early stages though, I couldn't tell you if it would be useful now.

**Comment by prnam** [comment]
## Search & Select Container Image feature

Objective 
-  Reduce the issues with wrong image pull or typo-sqautting attacks 
-  Make it even more easier to select image using container field present at create from form

Due date  - March 31, 2022

Key outcomes 
- Time taken to select image for users must reduce, thus improve user experience 
- Should minimise typo-sqautting attacks  
- Should widely be accepted by the community as stable feature for at least 6 months from launch

----
### 🤔 Problem Statement

Manual entry of container image name or digest without autocomplete may lead to unintended images pulls or typo-squatting attacks. Listing images from well known registries as in when user types in container image field present at create from form tab and making them able to select it if they are happy with suggestion will help reduce the issue and may lead more user adopt to the create from form page i.e in other words increase better user experience to our dashboard users.

### 🎯 Scope

Must have:

- container image field in create from form should be able to search & select the image from configured registry

- only list public images which are classified / vetted as verified or official

- registry configuration on settings page

- out-of-the-box pre-configuration of well known registry that are OCI compliant

- fallback to text based input (as it is now) on container image field when none is selected from the list or images or from pre-pulled or locally built images

- GET latest image information from all registry configured, every time create from form is selected by user

Nice to have:

- list all public images that are available in any registry and remove limitation to verified and official images
- to use logos than using wordmark as prefix for uniquely identifying images with same name 
   (ex: using docker logo and image-name:tag as text than using docker/ngnix:latest / acr.io/ngnix:latest / gcr.io/ngnix:latest as plain text in suggestion dropdown list )
- extend the ability to search & selected from image:tag to also using image@digest

Not in scope:
- private hosted registry configuration on settings page
- preview metadata of image within container image field in create from form
- list locally built images

 ### 📒 Registries

Name | Method | Out-of-the-box configured or Pre-configured | Comments | Mode (Public / Private / Private & Public)
-- | -- | -- | -- | --
Docker | [API](https://docs.docker.com/registry/spec/api/#detail) | YES | add auth as secrets for private registry access | Private & Public
Azure | [API](https://docs.microsoft.com/en-us/rest/api/containerregistry/) | YES | add auth as secrets for private registry access | Private & Public
OCI | [API](https://docs.oracle.com/en-us/iaas/api/#/en/registry/20160918/) | YES | add auth as secrets for private registry access | Private & Public
GCP | [API](https://cloud.google.com/container-registry/docs/reference/docker-api) | YES | add auth as secrets for private registry access | Private & Public
AWS | [API](https://docs.aws.amazon.com/AmazonECR/latest/APIReference/Welcome.html) | YES | add auth as secrets for private registry access | Private & Public
Quay | [API](https://access.redhat.com/documentation/en-us/red_hat_quay/3.6/html/red_hat_quay_api_guide/index) | NO | API are not documented well at the moment | N/A

**Comment by prnam** [comment]
Hey @maciaszczykm - Please find my initial project plan for this feature. I am working on high fidelity design and  will share it as soon as I am done.

**Comment by maciaszczykm** [comment]
@prnam Sounds nice, you can continue with creating proof of concept. I'd start with adding support for one registry, i.e. Docker. Please open pull request sooner rather than later so we will be able to discuss technical details as this doesn't cover them.

**Comment by prnam** [comment]
You read my mind, @maciaszczykm. Every weekend you will get the updates both design and technical with draft PR. Will tag you here or Slack, when done or need help. 🙂

**Comment by prnam** [comment]
Updates from my notes (so far) :- 
- [Postman Workspace](https://www.postman.com/gold-crater-431796/workspace/511) consisting of REST API that will be used to fetch Image list & tags
- Here is how to get pre-pulled / built images list -
  ```zsh
  curl --unix-socket /var/run/docker.sock http://localhost/images/json
  ```
- Here is how to configure Private registry
  - [Create Secret of registry](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/#create-a-secret-by-providing-credentials-on-the-command-line)
  - [Create Pod using Secret](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/#create-a-pod-that-uses-your-secret)

Stuck at & Help needed -

- How to [list catalogs](https://docs.docker.com/registry/spec/api/#listing-repositories) from DockerHub, GCR, OCIR and ECR as anonymous user via REST API
- How to list specific image and tags for given namespace ([registry-alias in AWS](https://docs.aws.amazon.com/AmazonECRPublic/latest/APIReference/Welcome.html)) & repository-name via REST API for [ECR](https://gallery.ecr.aws) & [OCIR](https://container-registry.oracle.com)
- How to only list OR sort the list by [verified images](https://gallery.ecr.aws/?verified=verified&page=1) / [verified publisher](https://hub.docker.com/search?q=&type=image&image_filter=store) / [verified accounts](https://gallery.ecr.aws/?verified=verified&page=1) for any public registry

**Comment by prnam** [comment]
Update:
- w.r.t. AWS Public Registry, I have raised help at [AWS:Repost](https://www.repost.aws/questions/QUFbQnGajYS8en6uYFhaKpTg/how-to-get-list-of-tags-using-registry-alias-repository-for-public-ecr). Hoping to get help some day.
- w.r.t. Docker Registry, received some help at [Slack -Docker Community](https://dockercommunity.slack.com/archives/C7GKACWDV/p1643630390043729)

Next Steps( moving ahead) :
- creating a high fidelity design on the possible approach and get reviewed
- once approach is approved based on design, raise initial PR with same approach from technical standpoint

Other:
-  K8s #sig-ui Slack discussion thread [here](https://kubernetes.slack.com/archives/C09R1JVU6/p1641237947039700)

**Comment by prnam** [comment]
I was wondering if we can use etcd to store temporary key-value data of the image from all known registry and thus allow search/auto complete? Would look into these deeper but putting it out know if anyone had this thoughts and would like to share if there is any success / failure stories?

**Comment by ashutosh887** [comment]
I would like to work on this @bryk @codegoblins

**Comment by ashutosh887** [comment]
Thanks @maciaszczykm 
Would you like me to recommend how to approach this issue?

**Comment by ashutosh887** [comment]
@maciaszczykm Can you please guide me a bit on this?
There's a lot of conversation in the thread that's confusing me

**Comment by maciaszczykm** [comment]
@ashutosh887 Sorry for no response. It's quite old issue and I would recommend checking again for possibilities that are out there, i.e. any libraries/APIs that would allow this.
